nupic/research/frameworks/dendrites/routing/hardcoded.py

```diff
@@ -0,0 +1,140 @@
+
+    :param d_in: the number of dimensions in the input to the routing function and test
+                 module
+    :param d_out: the number of dimensions in the sparse linear output of the routing
+                  function and test network
+    :param k: the number of unique random binary vectors in the routing function that
+              can "route" the sparse linear output, and also the number of unique
+              context vectors
+    :param d_context: the number of dimensions in the context vectors
+    :param dendrite_module: a torch.nn.Module subclass that implements a dendrite
+                            module in addition to a linear feed-forward module
+    :param context_weights_fn: a function that returns a 3D torch Tensor that gives the
+                               near-optimal dendrite values for the specified
+                               dendrite_module, and has parameters `output_masks`,
+                               `context_vectors`, and `num_dendrites`
+    :param batch_size: the number of test inputs
+    """
+
+    # Initialize routing function that this task will try to hardcode
+    r = RoutingFunction(d_in=d_in, d_out=d_out, k=k, sparsity=0.7)
+
+    # Initialize context vectors, where each context vector corresponds to an output
+    # mask in the routing function
+    context_vectors = generate_context_vectors(
+        num_contexts=k,
+        n_dim=d_context,
+        percent_on=0.2
+    )
+
+    # Initialize dendrite module using the same feed-forward sparse weights as the
+    # routing function; also note that the value passed to `dendrite_sparsity` is
+    # irrelevant since the context weights are subsequently overwritten
+    dendritic_network = dendrite_module(
+        module=r.sparse_weights.module,
+        num_segments=k,
+        dim_context=d_context,
+        module_sparsity=0.7,
+        dendrite_sparsity=0.0
+    )
+
+    dendritic_network.register_buffer(
+        "zero_mask",
+        copy.deepcopy(r.sparse_weights.zero_mask.half())
+    )
+
+    # Choose the context weights specifically so that they can gate the outputs of the
+    # forward module
+    if context_weights_fn is not None:
+        dendritic_network.segments.weights.data = context_weights_fn(
+            output_masks=r.output_masks, context_vectors=context_vectors,
+            num_dendrites=k
+        )
+
+    # Sample a random batch of inputs and random batch of context vectors, and perform
+    # hardcoded routing test
+    x_test = 4.0 * torch.rand((batch_size, d_in)) - 2.0  # sampled i.i.d. from U[-2, 2)
+    context_inds_test = randint(low=0, high=k, size=batch_size).tolist()
+    context_test = torch.stack(
+        [context_vectors[j, :] for j in context_inds_test], dim=0
+    )
+
+    target = r(context_inds_test, x_test)
+    actual = dendritic_network(x_test, context_test)
+
+    # Report mean absolute error
+    mean_abs_error = torch.abs(target - actual).mean().item()
```

Comment:
Return mean_abs_error from this function, and put the printouts in the main.

Reply:
Fixed.